lib/src/clojure_lsp/feature/restructure_keys.clj

```diff
@@ -0,0 +1,190 @@
+(ns clojure-lsp.feature.restructure-keys
+  (:require
+   [clojure-lsp.queries :as q]
+   [clojure-lsp.refactor.edit :as edit]
+   [clojure-lsp.shared :as shared]
+   [rewrite-clj.node :as n]
+   [rewrite-clj.zip :as z]))
+
+(defn ^:private map-to-restructure [zloc uri db]
+  (when zloc
+    (let [filename (shared/uri->filename uri)
+          {:keys [row col]} (meta (z/node zloc))
+          def-elem (q/find-definition-from-cursor db filename row col)
+          loc (or (when (and def-elem
+                             (= :locals (:bucket def-elem)))
+                    (when-let [up-loc (-> zloc
+                                          edit/to-top
+                                          (edit/find-at-pos (:row def-elem) (:col def-elem))
+                                          z/up)]
+                      (cond-> up-loc
+                        (= :vector (z/tag up-loc)) z/up)))
+                  zloc)]
+      (when (seq (q/find-locals-under-form db filename (meta (z/node loc))))
+        (cond
+          (z/map? loc) {:filename filename
+                        :replace-loc loc
+                        :map-loc loc}
+          (z/namespaced-map? loc) (let [qualifier-loc (z/down loc)
+                                        {:keys [auto-resolved? prefix]} (z/node qualifier-loc)]
+                                    {:filename filename
+                                     :replace-loc loc
+                                     :map-loc (z/right qualifier-loc)
+                                     :map-auto-resolved? auto-resolved?
+                                     :map-prefix prefix})
+          :else nil)))))
+
+(defn can-restructure-keys? [zloc uri db]
+  (boolean (map-to-restructure zloc uri db)))
```

Comment:
I wonder how much time this adds to code actions processing, thanks for adding that too BTW, my point is we could have a simple/faster check for code actions not handling complex cases and only fail during code action apply if necessary, only if this code proves affect performance of code actions. WDYT?

Reply:
So you're saying [this](https://github.com/clojure-lsp/clojure-lsp/blob/d611822eec77079a805168afb0cbb2c92ff626c4/lib/src/clojure_lsp/feature/code_actions.clj#L277-L299) worries you? Lol, me too!

IIRC, this feature didn't change the performance of code-actions noticeably. But I can plan to benchmark code-actions anyway. It's a bit tricky because it happens across so many threads, and the work that gets done varies a lot depending on your location. I'll see if I can identify restructure-keys as a performance problem.

If it is, you're right... the answer is to find a simpler code action check that then fails during command execution. I'd feel better about that if we showed the user an error about why we couldn't perform the refactoring. If you return a JSON-RPC error like [here](https://github.com/clojure-lsp/clojure-lsp/blob/d611822eec77079a805168afb0cbb2c92ff626c4/lib/src/clojure_lsp/feature/rename.clj#L231-L233) does the client show that? Or do you have to use `window/showMessage`?

Reply:
@ericdallo I didn't set up a flamegraph, but by watching the log, the code-actions performance doesn't seem to be a problem. It tends to range between 10ms to 40ms, or even sometimes 60ms. Both ends of that range feel fast enough.

The fact that `restructure-keys` can be included on the low end of the range suggests it's not contributing to the performance significantly. Anecdotally, it seems like the overall time is proportional to the size of the file. I'd guess that means rewrite-clj dominates the time, as it often does. So, that's something to keep in mind... if code-actions ever does get slow, we could try caching the rewrite-clj parse, as we've talked about before.

I'm going to go ahead and merge this.

Reply:
> If you return a JSON-RPC error like [here](https://github.com/clojure-lsp/clojure-lsp/blob/d611822eec77079a805168afb0cbb2c92ff626c4/lib/src/clojure_lsp/feature/rename.clj#L231-L233) does the client show that?

Not sure about other clients, but emacs would print the error IIRC

> The fact that restructure-keys can be included on the low end of the range suggests it's not contributing to the performance significantly. Anecdotally, it seems like the overall time is proportional to the size of the file. I'd guess that means rewrite-clj dominates the time, as it often does. So, that's something to keep in mind... if code-actions ever does get slow, we could try caching the rewrite-clj parse, as we've talked about before.

Exactly, I already spent some time tunning performance of those actions so it's expected to not be a issue right now, but we don't have control if we add a new code action that will bring all code action time because of that refactor, that was my concern, I don't believe this is the case for this action but we need to be aware of this

Reply:
Yeah, understood. This one was particularly good to be wary of because there are a few queries, not just traversals of the parse tree. Something else to add to the performance tests!